Rewrite Solutions management

gdemonet opened the pull request:
**Components**: salt, solutions, scripts

**Context**: Improvement of Solutions management UX

---

**To do**:

- [ ] Rework packaging of the CLI (no need to make it redistributable)
- [ ] Rework logging with verbosity toggle (need to think about more than two verbosity levels, if needed in the future)
- [x] Remove `Environment` CRD:
    - [x] Replace with `Namespace`(s) metadata:
        - Label `solutions.metalk8s.scality.com/environment=<environment_name>`
        - Annotation `solutions.metalk8s.scality.com/environment-description=<short_desc>`
    - [x] Adapt MetalK8s UI:
        - List environments based on the label (group namespaces having the same env name)
        - Create such a namespace instead of an `Environment` CR when asked to
        - Add/update ConfigMap describing deployed versions of Solution Operators in that namespace
- [x] Expose available versions to Admin UIs through ConfigMap (even if Salt doesn't use it for state management, it updates it for consumption)
- [x] Rework Solution images repository handling
    - [x] Rework configuration of Operators (need to have some config file, not just env var)
    - [ ] Rework configuration of Admin UI (similar to operator, to know which repositories exist for a Solution)
- [ ] Document how Operators/UIs should be defined so they can still work outside of MetalK8s (maybe requiring some manual Operations for now)

---

**Summary**:

### Part I - Rework Salt implementation

_The previous approach relied on some custom state management, where Salt uses a `ConfigMap` to store a list of available Solution versions, to compare with the desired state configured in `/etc/metalk8s/solutions.yaml`._

- Rework configuration file format to let users declare which Solution versions are *active* (meaning, their Admin UIs / CRDs should be deployed)
- Rework ext_pillar `metalk8s:solutions` to list *available* Solutions in the cluster, based on simple observations (active mounts, existing UI Service objects in `metalk8s-solutions` Namespace...) of the actual state, instead of the `ConfigMap` used before
- Use a single formula (`metalk8s.solutions.available`) to (un)mount archives and (un)configure the cluster OCI registry
- Use a single orchestrate (`metalk8s.orchestrate.solutions.deploy-components`) to (un)deploy Solution _components_ (Admin UI, CRDs)

### Part II - Rewrite `solution-manager.sh` in Python

_To use the new states and config file format, and provide a comprehensive user experience, we rewrite the script in Python with `argparse`._

- Define the skeleton for a `metalk8s` CLI entrypoint, using git-style subcommands (implemented with `argparse`)
- Add logging to match existing behaviour in Shell scripts
- Add Salt helpers for running commands on either the Salt Minion or the Salt Master
- Add subcommands to manage Solutions:
    - `/path/to/metalk8s solutions add /root/some-solution.iso`
    - `/path/to/metalk8s solutions remove /root/some-solution.iso`
    - `/path/to/metalk8s solutions deploy some-solution --latest`
    - `/path/to/metalk8s solutions deploy some-solution --use-version 1.2.3-alpha1`
    - `/path/to/metalk8s solutions deploy some-solution --delete`

**Acceptance criteria**:

- Bootstrap still works, and after it's done, all Solution dependencies are deployed:
    - `metalk8s-solutions` Namespace exists
    - `ui-branding` ConfigMap exists in `metalk8s-solutions` namespace
    - _Environment_ CRD exists (`environments.solutions.metalk8s.scality.com`)

- One can add a Solution archive to the cluster (see #1876 for how to build the example ISO):
    - `/srv/scality/metalk8s-2.4.0-dev/metalk8s solutions add /vagrant/.../example-solution-0.1.0-dev.iso`
    - After this is done, one can retrieve container images from this ISO: `crictl pull -D metalk8s-registry-from-config.invalid/example-solution-0.1.0-dev/base-server:0.1.0-dev` should work fine
    - `salt-call pillar.get metalk8s:solutions:config:archives` should contain `/vagrant/.../example-solution-0.1.0-dev.iso`
    - `salt-call pillar.get metalk8s:solutions:available:example-solution` should show the relevant information about this archive and its mountpoint, and `active` will be `False`

- One can deploy components for a Solution version:
    - `/srv/scality/metalk8s-2.4.0-dev/metalk8s solutions deploy example-solution --latest`
    - After this is done, one can access the Solution UI (`kubectl get svc -n metalk8s-solutions example-solution-ui` to retrieve the UI port, its IP should be the workload-plane IP)

---

Fixes: #1582
Fixes: #1614
Fixes: #1852 
Fixes: #1854

See: #1588

gdemonet committed:
salt/solutions: Revamp state for available ISOs

Making a Solution archive "available" to the cluster was split into two
formulas: `metalk8s.solutions.mounted` and `.configured`.
We group them into a single one, called `.available`.

We also remove the `.unmounted` and `.unconfigured` formulas, since the
source of truth is entirely held by the configuration file
(/etc/metalk8s/solutions.yaml). A single formula is responsible for
applying the desired state described in this configuration, adding and
removing Solution archives from the cluster.

gdemonet committed:
salt/solutions: Deploy Solutions K8s dependencies

Some objects are required for MetalK8s to manage Solutions, notably a
Namespace and ConfigMap for Admin UIs, and the `Environment` CRD.

Issue: #1852

gdemonet committed:
salt/solutions: Revamp configuration file

A new field, currently named `active` (may be a poor choice, we'll see
with first user interactions, `activeVersions` could be more explicit),
is introduced to set which version of a Solution should be used for the
cluster-wide components of this Solution (i.e. the Admin UI and CRDs for
now).
Validation of the file format is added, using the usual `kind` and
`apiVersion` fields. Later improvements could use some OpenAPI schema.

Fixes: #1582

gdemonet committed:
salt/solutions: Revamp current state in pillar

Previous approached relied on a ConfigMap, which poses numerous issues:
- it allows one to lie to the formula, potentially causing further
problems down the line
- the CM cannot always be up-to-date with the actual system state (due
to errors when updating it, or manual operations outside of Salt)

We change this approach by including basic heuristics in the
`metalk8s_solutions` module, which can be refined in the future.

Listing of mounted Solutions is made based on `mount.available` and a
naive filtering on the mountpoints (necessarily under /srv/scality).

Listing of active Solution versions is made by looking for UI Service
objects in the `metalk8s-solutions` namespace. This approach relies on
labels that we don't set ourselves, which is brittle, and does not
account for Solutions that may want to ship without an Admin UI.

Future iterations could improve those methods, but we are convinced it
is still safer than the ConfigMap method used before.

Issue: #1852

gdemonet committed:
salt/solutions: Revamp orchestrate for K8s objects

Once a Solution is "available" (i.e. mounted and its images exposed by
the cluster registry), its cluster-wide components can be deployed.

We rely on the `active` mapping of active versions in the config file to
know which components to deploy or remove.

Note that one can set a desired version to 'latest' in this mapping,
which the orchestration will handle for using the highest version
available for a Solution (in SemVer terms).

Issue: #1852

gdemonet committed:
scripts: Init metalk8s_cli Python implementation

To ease future improvements to our CLI scripts and increase robustness
of these execution, we want to rewrite these scripts in Python.
This commit introduces a dummy `metalk8s` Bash entrypoint to execute the
`metalk8s_cli` Python package, based on `argparse`.

See: #1588
See: #1854

gdemonet committed:
scripts: Monkey-patch old versions of argparse

Old (<2.7.9) versions of `argparse` have a different behaviour when it
comes to defaults set by subparsers. We need the most recent behaviour
for injecting the selected `Command` class.

Alternatives to this monkey-patch were deemed too complex in comparison
(e.g. enforcing Python 3, with the implications in terms of packaging
for CentOS/RHEL 7).

gdemonet committed:
scripts: Add subcommand for Solutions management

In order to properly isolate commands in our CLI, we use a dedicated
namespace for Solutions management. These commands will be invoked
through "metalk8s solutions ...".

Since most of these commands require access to the
SolutionsConfiguration file, we create a helper class to manage this
file contents.

Issue: #1854

gdemonet committed:
scripts: Add logging support

For our CLI, we want to be able to adjust verbosity of stdout messages,
as well as store as much information as possible in logfiles.

To reproduce this behaviour and provide loggers to `Command` sub-classes,
we define a `LoggingCommandMixin` and some global CLI args (`--verbose`,
`--logfile`).

Issue: #1588
Issue: #1854

gdemonet committed:
scripts: Add mixin for Salt-related commands

To help write more readable commands, we provide numerous helpers for
interacting with Salt through its own CLIs.

Current implementation is based on `subprocess`.
Future improvements could use the Python client and SaltAPI instead.

Note that all of this dramatically lacks a performance optimization
pass, there are some low-hanging fruits, on properties caching for
instance.

See: #1588
See: #1854

gdemonet committed:
scripts: Define commands to add/remove Solutions

These commands are simple: first, edit the config file from user input,
then run `salt-call state.sls metalk8s.solutions.available`.

Fixes: #1614
Issue: #1852
Issue: #1854

gdemonet committed:
scripts: Add command to (de)activate a Solution

In that sense, activating a Solution version is merely setting this
version in the config file and running the
`metalk8s.orchestrate.solutions.deploy-components` orchestration.

Users can either specify the desired version by hand, using
`--use-version`, or choose the latest one, using `--latest`.

Users can also deactivate a Solution (removing its Admin UI and CRDs) by
passing the `-d`/`--delete` flag.

Issue: #1852
Issue: #1854

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented:
Forgot that I need to adjust the UIs to work without the ConfigMap... For reviewers, please don't focus on UIs right now.

TeddyAndrieux commented on salt/_modules/metalk8s_solutions_k8s.py:
```diff
@@ -0,0 +1,71 @@
+            solution_name = labels["app.kubernetes.io/part-of"]
+            solution_version = labels["app.kubernetes.io/version"]
+        except KeyError:
+            # FIXME: ignoring invalid Service objects for now, though it may
```
At least warning IMO

gdemonet replied:
Good idea, will do

TeddyAndrieux commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -1,186 +1,122 @@
+    Empty containers will be used for `archives` and `active` in the return
+    value.
+
+    The format is expected to resemble the following example:
```
```suggestion
    The format should look like the following example:
```

TeddyAndrieux commented on salt/metalk8s/solutions/available.sls:
```diff
@@ -0,0 +1,120 @@
+        registry_root: {{ mount_path }}/images
+    - require:
+      - file: Container images for Solution {{ display_name }} exist
+    - onchanges_in:
```
Would suggest `require_in` because `onchanges_in` that mean if for one reason only one `Expose container images for Solution blabla` fail you will not run the `metalk8s.repo.installed` state and if you run again and no changes made you will not run the state. (This sentence does not look really clear)

Example:
You want to make toto and tata solutions available:
- toto not configured at all
- tata already has a good `registry-config.inc` but for what ever reason product.txt not here

1- Launch this state => toto get configured with good `registry-config.inc` tata fail => `metalk8s.repo.installed` is not executed because one "onchanges" state failed
2- Fix tata product.txt
3- Launch this state again => All green nothing change (repo still not configured)

With a `require_in` the 3- will run the `metalk8s.repo.installed` state.

(For me `onchanges` it's something only usefull in really specific case for state that is not idempotent and make "dangerous" changes)

gdemonet replied:
OK I get the reasoning, thanks for the explanation! Will do :)

TeddyAndrieux commented on salt/metalk8s/solutions/available.sls:
```diff
@@ -0,0 +1,120 @@
+{# Validate the archive contents
+   TODO: This should be moved before mounting the solution's ISO, using some
+   custom module #}
+Product information for Solution {{ display_name }} exists:
```
If you try to make 2 different version of the same solution at the same time you will have conflict here as you will have 2 states with the same name.

You should change the display_name to also contains the version or add something unique for each state so "display_name" + "version" or "mount_path" but not "display_name" only

gdemonet replied:
See L.29: `{% set display_name = solution.name ~ ' ' ~ solution.version %}`

TeddyAndrieux replied:
Oh yes MB through it was the `display_name` from `list_available` salt modules

TeddyAndrieux commented on salt/metalk8s/solutions/available.sls:
```diff
@@ -0,0 +1,120 @@
+      {%- if info.archive not in configured %}
+        {%- set display_name = info.display_name ~ ' ' ~ info.version %}
+        {%- if info.active %}
+Cannot remove archive for active Solution {{ display_name }}:
```
Ditto for remove you need to have a unique state name or if you remove 2 different versions of the same solution the state will fail

TeddyAndrieux commented on salt/metalk8s/solutions/available.sls:
```diff
@@ -0,0 +1,120 @@
+Remove container images for Solution {{ display_name }}:
+  file.absent:
+    - name: {{ repo.config.directory }}/{{ name }}-registry-config.inc
+    - onchanges_in:
```
ditto :)

TeddyAndrieux commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -1,186 +1,122 @@
+
+        result[machine_name].append({
+            'display_name': solution_info['name'],
+            'machine_id': '{}-{}'.format(machine_name, version),
```
`solution_id` may be better no ? Because nothing really related to "machine" here right ?

gdemonet replied:
The point is this name/ID is for machine use, while `display_name` is intended for humans. Maybe I can add a comment to explicit this distinction.

TeddyAndrieux replied:
Oh okay machine in that sense

slaperche-scality replied:
Maybe simply call them `name` and `id`, as `name` are usually for human whereas `id` are usually for computer.

TeddyAndrieux reviewed (changes requested):
First quick review (only for salt part, will do another one for scripts :) )

But looks pretty good

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

TeddyAndrieux commented on scripts/metalk8s:
```diff
@@ -0,0 +1,45 @@
+
+pushd "$(dirname ${BASH_SOURCE[0]})" > /dev/null
+SCRIPTS_DIR="$(pwd)"
+popd > /dev/null
```
Hmm personnaly prefer somethink like that 
```
SCRIPTS_DIR="$(dirname "$(readlink -f "$0")")"
```
but maybe `pushd` `popd` better I'm not a bash expert @alexandre-allard-scality ? :)

gdemonet replied:
I stole this whole Bash script from @NicolasT ;)
But I agree your one-liner looks more explicit to me, being a noob at Shell scripting.

TeddyAndrieux commented on scripts/metalk8s:
```diff
@@ -0,0 +1,45 @@
+SCRIPTS_DIR="$(pwd)"
+popd > /dev/null
+
+if [ ! -d "${SCRIPTS_DIR}" ]; then
```
Wut ? It's possible ?

gdemonet replied:
:man_shrugging: I'll remove the check I guess

TeddyAndrieux commented on scripts/metalk8s_cli/parsers.py:
```diff
@@ -0,0 +1,30 @@
+def build_logfile_parser():
+    """Allow defining a logfile for this command execution."""
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument('--logfile',
```
`-l` also

gdemonet replied:
OK

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/log.py:
```diff
@@ -0,0 +1,251 @@
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        open(self.logfile, 'a').close()
```
Does this really needed ?

gdemonet replied:
Indeed it's not required, file will get created at first log (so should be the `self._init_run_log` call actually). Will remove it :)

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/log.py:
```diff
@@ -0,0 +1,251 @@
+    def __init__(self, args):
+        super(LoggingCommandMixin, self).__init__(args)
+        self._logger = None
+        self.logfile = os.path.abspath(args.logfile)
```
For me we should do like in the ring installer with one global log file and one log file specific for each steps :)

So if you run bootstrap each log are in /var/log/metalk8s/bootstrap.log and /var/log/metalk8s/metalk8s.log, if you run upgrade each log are in /var/log/metalk8s/upgrade.log and /var/log/metalk8s/metalk8s.log ... (but maybe not part of this issue/PR IMO)

gdemonet replied:
Good suggestion! I was not sure about splitting command logs, but having a _catch-all_ logfile in addition is smart :)

I'll do it in this PR, it's not much of a change anyway (iff we agree that `/var/log/metalk8s/metalk8s.log` is a non-configurable value for the catch-all logfile).

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/log.py:
```diff
@@ -0,0 +1,251 @@
+        finally:
+            self._stop_run_log()
+
+    def print_and_log(self, message, level=logging.INFO, keep_line=False):
```
Hmm it's weird to call this function each time you want to log something.

For me this class should inherit from classic Logging class and dynamicaly change the level of the console handler when verbose mode enable

gdemonet replied:
> this class should inherit from classic Logging class

WDYM?

> dynamicaly change the level of the console handler when verbose mode enable

Huh, I did this first, then decided otherwise. The rationale is the following: it's a CLI, so there's a big difference between what goes in `stdout` and `stderr`. IMO (probably wrong), _verbosity_ of the script affects what goes in `stdout`, while log messages go in `stderr` (and we may give knobs to configure this logging for users). This way, if commands are supposed to return parsable output, a consumer can easily redirect `stderr` and avoid any risk when parsing (as you mentioned for the `SaltCommandMixin._get_json_from_minion` :wink:).

Now, I wanted to have the `print`-ed messages in the logfile as well, which is why there's this `print_and_log` method.

> it's weird to call this function each time you want to log something.

I think some occasions may justify using "log-only" messages for `INFO` level and below, but for now all of those were user-oriented. Note that errors are not `print`-ed.

TeddyAndrieux replied:
> WDYM?

Why not create a `class Metalk8sLogger(logging.getLoggerClass())` ?
Then we could use directly the classic `debug`, `info` ...

> I think some occasions may justify using "log-only" messages for `INFO` level and below, but for now all of those were user-oriented. Note that errors are not `print`-ed.

For me the only things we want to print is the name of the step + result "Deploying bootstrap node (this may take a while)... done[200s]" others stuff may only go in log and if you enable verbose mode then you change the handler for console so that log also printed in the console.

So IMO printing of the beginning of the step + result should be done in `log_step` function directly. Instead of `print_and_log` we should `print` and `log` (maybe something with more detail if needed)

(Btw it's only my opinion feel free to dismiss this one if you want :) )

gdemonet replied:
> Why not create a class `Metalk8sLogger(logging.getLoggerClass())` ?
Then we could use directly the classic `debug`, `info` ...

Why would we need a custom logger? Isn't the native one sufficient? If you want the `LoggingCommandMixin` to **be** a `Logger` subclass, then I disagree, we should favor composition in that kind of situation.

> if you enable verbose mode then you change the handler for console so that log also printed in the console

See above as to why I don't want to do this ;)

> printing of the beginning of the step + result should be done in `log_step` function directly. Instead of `print_and_log` we should `print` and `log` (maybe something with more detail if needed)

OK, I think that's a good middle-ground. I'll need to filter out `INFO` messages from the `console` handler to avoid duplication, but OK.

TeddyAndrieux replied:
> Why would we need a custom logger? Isn't the native one sufficient? If you want the `LoggingCommandMixin` to **be** a `Logger` subclass, then I disagree, we should favor composition in that kind of situation.

Not really needed but if we want to have some special feature this can help but for the moment maybe not needed.

> > if you enable verbose mode then you change the handler for console so that log also printed in the console
> 
> See above as to why I don't want to do this ;)

Which one ?
You can change the log level of an handler without having a subclass of `Logger`

> OK, I think that's a good middle-ground. I'll need to filter out `INFO` messages from the `console` handler to avoid duplication, but OK.

Why ? We should have no duplication here if you never print in a logging function but only log

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/salt.py:
```diff
@@ -0,0 +1,119 @@
+        self, command, local=False, outputter=None, color=True,
+        saltenv=None, pillar=None,
+    ):
+        full_cmd = ['salt-call']
```
`--retcode-passthrough` I think even if we don't want to fail I think it's good to have the return code

gdemonet replied:
Good point, will do (those helpers are really a draft, so thanks for the help!)

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/salt.py:
```diff
@@ -0,0 +1,119 @@
+        try:
+            return subprocess.check_output(full_cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as exc:
+            raise CommandError(
```
Do we really want to raise here ? For me we may want some time to do something with the output in case of error.

For me we should either:
1- always return a dict with `stdout` `stderr` `retcode`
2- when `retcode != 0` raise something that contain `stderr` `stdout` and `retcode` (I mean a dict containing these informations)

For me 2- look like a good solution most of the time we will just not catch the exception and will get the msg just here but for rare special case we can `try except` and get the stderr stdout, retcode and do what ever we want with this

gdemonet replied:
Indeed, I was thinking of implementing the second approach. Using `check_output(cmd, stderr=PIPE)` should do the trick.

gdemonet replied:
Now there's a `_run_process` helper, which gives you a `CompletedProcess` namedtuple, with keys `args`, `returncode`, `stdout` and `stderr`. In case `returncode` was not 0, it instead raises a `CommandSubprocessError`, with the same attributes, for callers to catch and log as they want.

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/salt.py:
```diff
@@ -0,0 +1,119 @@
+            )
+
+    def run_salt_master(self, command):
+        # TODO: add cache for the salt-master container ID
```
I don't know if we can/want really to have cache for this.
Because if we cache this we need either
1- a variable to "force caching again" 
2- something to check that the container ID still good (which is basically the command executed in `get_salt_master_container`)

gdemonet replied:
Indeed, I was actually thinking about giving control to the caller (same as for `saltenv` in `run_salt_minion`, callers store it once and can pass it to each call). This way, if a command expects the container ID to have changed, it can refresh its value, while otherwise saving the cost of running the subprocess again.

TeddyAndrieux commented on scripts/metalk8s_cli/mixins/salt.py:
```diff
@@ -0,0 +1,119 @@
+        out_string = self.run_salt_minion(
+            command, outputter='json', color=False, **kwargs
+        )
+        return json.loads(out_string)["local"]
```
Care here since you do `stderr=subprocess.STDOUT` in subprocess call if for what ever reason salt log something to stderr you will likely not be able to parse this as json.

gdemonet replied:
Indeed, will change. I think I'll remove `check_output` and use `run` instead, that will allow more control for callers.

slaperche-scality replied:
`subprocess.run` was introduced in Python 3.5, dunno if it's "old enough" to be used in this context.

gdemonet replied:
Again... OK so I'll do another monkey-patch of this `subprocess.run` as we do for `argparse`, the ugliness of these should encourage us in only using Python 3 :)

slaperche-scality replied:
I mean, it's our custom script so maybe we can use Python 3, but even then I'm not sure CentOS 7 comes with Python 3.5+ (it's Python 3.4, no?)

gdemonet replied:
There's no Python 3 pre-installed on CentOS 7. There's 3.6 in `base`, though most of the usual dependencies (e.g. `six` or `PyYAML`) are not, so we'd need to include them in our archive, and somehow install them before execution of the bootstrap script (since it's intended to run through this Python CLI).

I think it could be feasible, but would cost more efforts for us. So let's go with something simpler for now :) and once everything is moved to Python 3 (Salt, notably), then we'll be able to enforce it for the script as well.

gdemonet replied:
Mmh, monkey-patch is not that easy, there's quite a lot of changes in the `run` API actually, when compared to the capabilities of `subprocess` in 2.7... So I'll ignore `run` and only expose `stderr` in case of error.

TeddyAndrieux replied:
Why do not use directly `subprocess.Popen` ?

gdemonet replied:
`run` provides a bit more flexibility, so I made a _poor man's_ variant of it using `Popen.communicate`. It lacks `KeyboardInterrupt` or timeout management though, but should be fine for a first version.

slaperche-scality replied:
What's the deal with `KeyboardInterrupt`?

For timeout indeed, there is nothing in Python 2…

gdemonet replied:
So, `run` in Python 3 relies on using `Popen` as a contextmanager, which performs clean-up of open streams and handles `KeyboardInterrupt` errors when doing so. I just ignored the clean-up of stdin (which may raise other errors), as it's not in our use-cases for now. See [`subprocess.Popen.__exit__`](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L998-L1024)

TeddyAndrieux commented on scripts/metalk8s_cli/solutions/config.py:
```diff
@@ -0,0 +1,123 @@
+        )
+        assert api_version in self.API_VERSIONS, (
+            'API version not supported: {}'.format(api_version)
+        )
```
Hmm you do the same check as in the salt module. That mean if for what ever reason we need to change something we need to change it in salt and in the script. Maybe you should use a salt module to check this config file.

gdemonet replied:
Good point. Given it's a simple `salt-call --local`, I guess the performance penalty of delegating this to Salt is acceptable (this CLI is already slow as hell due to the numerous calls to Salt master - might be good to profile a bit more what takes time, I fear having expensive ext_pillars is not that good for responsiveness of the system).

gdemonet replied:
Also, since Salt is responsible for defining format/usage of this config file (same as it would be for `/etc/metalk8s/bootstrap.yaml` I guess), I think it should also be responsible for managing its path (for now, it's hardcoded in `__salt__['metalk8s_solutions.SOLUTIONS_CONFIG_FILE']`). Which means I should remove the `--solutions-config` argument.

gdemonet replied:
Done, see 65d8738

TeddyAndrieux reviewed (changes requested):
Several comments (not tested for the moment) but looks really better than our current bash scripts gj

gdemonet committed:
salt/solutions: Rename some attributes in pillar

We used `display_name` and `machine_id` to differenciate the expected
consumers (user or machine). For simplicity (thanks to slaperche-scality),
we rename them in `name` and `id` respectively. KISS!

slaperche-scality commented on scripts/metalk8s_cli/mixins/salt.py:
```diff
@@ -0,0 +1,119 @@
+        try:
+            return subprocess.check_output(
+                # FIXME: pillar args... need to escape?
+                ' '.join(full_cmd), shell=True,
```
hum yeah, this looks dangerous, passing a string coming from non-controlled input (pillar) and let the shell interpret it, with no escaping.

Why not use a list of string here?
And if you absolutely want to use a string, then use `shlex.quote`.

Now, being Python 2 it's probably not available (otherwise it's not fun), but IIRC the implementation is not that complex (but still, would be better to use a list of string and keep `shell=False`).

gdemonet replied:
My bad, the `FIXME` comment was not explicit. The issue I was faced with when using a list of strings, the value `"pillar='{bootstrap_id: bootstrap}'"` is not well interpreted by Salt, and fails with some `invalid pillar` exception. Maybe it's not well escaped when sent to the execution context, so I'll give a try to `shlex.quote` on this specific argument.

slaperche-scality replied:
Weird, the whole point of using a list of string is to avoid the need for escaping (because the values are directly passed to the process).

This works¹ for me:

```python
pillar = {"update_archive_name": "new_archive_name.zip"}
cmd = ['salt', '*', 'state.highstate', 'pillar', json.dumps(pillar)]
subprocess.run(cmd)
```

¹: works as in `doesn't complain about the pillar`, but it fails with `No Top file or master_tops data matches found` because my local salt is not configured :stuck_out_tongue:

gdemonet replied:
I did not have issues with direct usage of `salt-call` either, the issue arose when sending to Salt Master through `crictl exec`... Did not test yet if the `shlex.quote` approach helps, so we'll see :)

slaperche-scality replied:
Ha, you may need to wrap the pillar between single quote then.

I think we're doing something like that in the expansion test `tests/install/steps/test_expansion.py`, maybe you can get an idea there.

gdemonet replied:
Huh, it seems the tests are doing it with `shell=True` as well (using `host.run(some_format_string)`) :/ and through `kubectl`. Tried switching to `kubectl` as well, used different escaping strategies, nothing works. I'll go back to `shell=True` as well here, actually `pillar` is not provided by the user, so I'd say it's not really _unsafe_ to use it within the CLI code.

slaperche-scality replied:
Yeah indeed, the pillar value is hardcoded here so it should be safe: let's go for that.
(but it's weird that we can't make it works without this)

slaperche-scality commented on scripts/metalk8s_cli/mixins/salt.py:
```diff
@@ -0,0 +1,119 @@
+            full_cmd.append("pillar='{}'".format(json.dumps(pillar)))
+
+        try:
+            return subprocess.check_output(full_cmd, stderr=subprocess.STDOUT)
```
Why do we, at several places, redirect `stderr` into `stdout`?

Why not using `subprocess.Popen` and then use `communicate` to get the output (`retcode` will be in the object returned by `Popen`)?

gdemonet replied:
The goal was to print `stdout`/`stderr` if the user requests `--verbose` mode only, but as @TeddyAndrieux suggested, we should not mix them if we want to parse the `stdout` result.
I moved to `Popen.communicate` since then, push is coming soon

gdemonet committed:
scripts: Improve management of subprocess

The initial crude implementation relied on `check_output`. We now
implement a poor man's `subprocess.run`, with defaults suiting our use
case (namely, an equivalent of the `CalledProcessError` being raised for
loggers to catch and display, `stdout` and `stderr` being captured).
This will be applied to existing commands in a later commit.

NicolasT commented on buildchain/buildchain/iso.py:
```diff
@@ -107,7 +107,53 @@
         task_dep=['_iso_mkdir_root']
-    )
+    ),
+    helper.FileTree(
```
I wonder whether it wouldn't be better to turn `metalk8s_cli` into a proper Python package, `sdist` it during build into `_build/...` (not `_build/root`), then 'install' it in `_build/root/whatever` (also during build).

gdemonet replied:
I'll give it a try :)

gdemonet replied:
OK, I think it's worth a dedicated issue, so we can think about different technical solutions (and what we want to do in the future for this CLI). I'll open one

NicolasT commented on buildchain/buildchain/iso.py:
```diff
@@ -107,7 +107,53 @@
+        files=(
+            Path('metalk8s'),
+
+            Path('metalk8s_cli/__init__.py'),
```
Not sure this should be in the project root. Is this *the* CLI to work with MetalK8s? I.e., will it include the stuff we wanted to put in the `kubectl` plugin?

gdemonet replied:
I think we can make it _the_ CLI for MetalK8s, and as a Python package, export `entrypoints` for some commands we want to install/use as `kubectl` plugin(s). Would this be a problem?
I also don't see why it should not be an installable/shareable package, as `kubectl` plugins would be.

NicolasT commented on buildchain/buildchain/iso.py:
```diff
@@ -107,7 +107,53 @@
+
+            Path('metalk8s_cli/__init__.py'),
+            Path('metalk8s_cli/__main__.py'),
+            helper.TemplateFile(
```
Not sure we need this, can read it (at runtime) from `product.txt`, since this is not a redistributable/system-wide-installable package

gdemonet replied:
~OK, will do~ Well, depends on the discussion regarding `kubectl` plugins being included in this Python package (which may contain multiple entrypoints). It may make sense to know the sources version when they are copied outside of the ISO.

gdemonet replied:
Discussion concluded that we will not include `kubectl` plugin(s) in this package, as they would probably be written in Go to make use of the available libraries.

NicolasT commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -221,8 +221,10 @@ def _get_parts(self) -> Iterator[str]:
-    Path('salt/metalk8s/addons/solutions/deployed.sls'),
-    Path('salt/metalk8s/addons/solutions/environment-crd.sls'),
+    Path('salt/metalk8s/addons/solutions/deployed/init.sls'),
+    Path('salt/metalk8s/addons/solutions/deployed/environment-crd.sls'),
```
Oh, an *environment* is an actual CRD? I though it maps to a *Namespace*?

gdemonet replied:
No, an `Environment` is actually a "group" of `Namespace`s with information about which version of a Solution (Operator) is running in such `Namespace`s (one namespace per Solution, to keep things organized by default). So it's a CRD (for simplicity, although we may want to change this in the future), but it's not tied to any automatic action through a controller.

gdemonet replied:
This concept was introduced to help users unfamiliar with namespaces to correctly manage Solutions without using a single namespace for a bunch of Solutions (the UI enforces a single Operator per Solution in each Environment, so actually each Operator has its own Namespace)

NicolasT commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -453,16 +451,16 @@ def _get_parts(self) -> Iterator[str]:
     Path('salt/_modules/metalk8s_solutions.py'),
-
+    Path('salt/_modules/metalk8s_volumes.py'),
+    Path('salt/_modules/metalk8s.py'),
```
That's a surprising reordering.

gdemonet replied:
Indeed, my editor shows them in this order, but I prefer the "usual" lexicographical ordering (such as the one for Python strings). I'll change this.

NicolasT reviewed (changes requested):
Intermediate POST

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

gdemonet committed:
salt/solutions: Manage config file from Salt

Instead of having validation logic in two places, we only use Salt to
read/write the Solutions config file, and use this Salt module from the
`metalk8s` CLI scripts.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/rewrite-solution-management` and the
destination branch `development/2.4`.

Please resolve the conflict on **the feature branch** (`improvement/rewrite-solution-management`).

```bash
 $ git fetch
 $ git checkout origin/improvement/rewrite-solution-management
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/rewrite-solution-management
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

TeddyAndrieux commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -1,186 +1,203 @@
-    from urllib3.exceptions import HTTPError
-except ImportError:
-    HAS_LIBS = False
+import yaml
```
Why did you move `import yaml` ?
```
C: 11, 0: third party import "import yaml" should be placed before "from salt.exceptions import CommandExecutionError" (wrong-import-order)
```

gdemonet replied:
Wut?! As your linter (`pylint` I guess) says, `yaml` is third party, not standard lib. So I put them together, and since `Y > S`, I put it after `salt`. Now, maybe you have a `pylint` plugin that wants `salt` imports after those, but I'm quite unsure as to why that would be a good idea.

TeddyAndrieux replied:
Oh yes indeed I use saltpylint. 
https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#linting

So yes it's 
```
# import builtins

# import thirdparty

# import salt_stuff
```

TeddyAndrieux commented on scripts/metalk8s:
```diff
@@ -0,0 +1,39 @@
+set -o pipefail
+
+SCRIPT_PATH="$(readlink -f "${BASH_SOURCE[0]}")"
+SCRIPTS_DIR="$(cd "$(dirname "${SCRIPT_PATH}")" && pwd)"
```
Hmm why `cd` + `pwd` ? `dirname "$SCRIPT_PATH"` is not enough ?

gdemonet replied:
I think the rationale was "if `dirname ...` is not an actual directory, then it will fail early"

gdemonet replied:
But I can remove it if you'd like

TeddyAndrieux replied:
Hmm maybe simpler/cleaner to do `test -d` for this.
But I don't see a case where a `dirname` of something who come from a `readlink -f` can give a non-directory file :confused:

TeddyAndrieux commented on scripts/metalk8s_cli/solutions/config.py:
```diff
@@ -0,0 +1,51 @@
+"""Helper module for managing a SolutionsConfiguration file."""
+import argparse
+import os.path
+
+import yaml
+
+from metalk8s_cli.exceptions import CommandInitError, CommandError
```
Look like all these imports are no longer used here. Only salt is needed

TeddyAndrieux reviewed (changes requested):
Few comments about the code (not tested yet, but I will)

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/rewrite-solution-management` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/rewrite-solution-management`
* Rebase `improvement/rewrite-solution-management` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

gdemonet committed:
scripts/cli: Refactor logging framework

We extract most of the logging logic into the `metalk8s_cli.log` module.
We also introduce a `verbosity` parameter, along with a `log-level` one,
to allow more fine-grained control of the CLI output.
Commands should now rely on `self.print` to write to `stdout`, and
`self.log` or its variants to log information (which will be written to
both `stderr` and some configured files).

gdemonet committed:
scripts/cli: Handle command initialization errors

We introduced a `CommandInitError` for Command classes to raise during
instantiation, but didn't catch it.

gdemonet committed:
WIP: remove Environment CRD

gdemonet committed:
ui/api: Split K8s APIs by domain

To help with organization of the increasing number of API call helpers,
we split them into domain-based modules, e.g. `k8s/solutions` or
`k8s/core`.

gdemonet committed:
[WIP] salt/solutions: environments in Pillar

gdemonet committed:
[WIP} salt/solution: Reinstate ConfigMap

We re-instate the ConfigMap as both a way to track which version was
made available (through the `deploy-components` orchestrate) and a way
to share the information with MetalK8s UI users.

gdemonet committed:
[WIP] salt/solution: Stop deploying Admin UI cluster-wide

The setup of "environment" Namespaces and all the resources required to
run a Solution instance in it is becoming too complex for it to be
properly managed by the Solution Admin UIs, without mentioning the authz
requirements on users of said Admin UIs to make it work.
Instead, Admin UIs will now be deployed once per environment, making it
easier to keep in sync with the same Operator version.

gdemonet committed:
[WIP] salt/solution: Support SolutionConfig files

Solution ISOs can now contain a `config.yaml` file to describe the few
specifics of its Operator and UI. We expose this information in the
`metalk8s:solutions:available` pillar, for later consumption in
orchestrations.

gdemonet committed:
[WIP] salt/solution: Prepare environments

Salt is now able to "prepare" environments, based on two sources of
information:

- the environment *Namespace*s (denoted by some well-known labels) and
  their 'metalk8s-environment' *ConfigMap*s (which lists the Solution
  versions to deploy in this environment)
- the *SolutionConfig*s stored in each Solution ISO which allows Salt to
  know what should be deployed

Both sources of information are stored in the pillar.

gdemonet committed:
ui/solutions: Remove UI services from state

We used to consider Admin UIs as unique per Solution across a cluster.
That's however not the case anymore, so we remove this data from the
Redux state.

gdemonet committed:
ui/solutions: Rework environments API

The environments are not exposed using CustomObjects anymore.
Instead, we manage them through labels on namespaces.

gdemonet committed:
ui/solutions: Read Solutions info in environments

We add more data about configured Solutions in an environment (through a
'metalk8s-environment' ConfigMap), and the Admin UI for each.
This will be used in the Solutions page to display Environments info.

gdemonet committed:
ui/solutions: Simplify list of available Solutions

We do not need to retrieve Admin UIs in the list of configured Solutions
that we retrieve from the 'metalk8s-solutions' ConfigMap, since those
Admin UIs are now instantiated for each environment.

gdemonet committed:
ui/solutions: Highlight active Solution versions

This makes the "active" version of a Solution stand out from the others.
An improvement would be to show a tooltip for this column, explaining
that an "active" version is the one for which cluster-wide components
(such as storage classes or CRDs) are deployed.

gdemonet committed:
ui/solutions: Rework listing of solutions and envs

We update the display of Solutions and Environments tables based on the
new data available in the Redux state.

gdemonet committed:
[WIP] ui: Run app behind /metalk8s

Not sure if useful, though Solution UIs will need a more generic way to
make this work with different basename at runtime...

gdemonet committed:
ui/solutions: Rework refresh saga

Use a more explicit pattern for the refresh saga, based on `take`,
`fork` and `race`.

See: #2035

gdemonet committed:
WIP - refresh Solutions only when on Solutions page

gdemonet committed:
WIP - serve fonts CSS rules as part of brand/

gdemonet committed:
WIP - rework management of Salt Jobs

We used to manage them only for Node deployment. A more generic
solution is introduced, and decoupling of the actual job polling from
consumption of the job status and result is achieved by dedicating a
duck (`state.app.salt`) to the job management.

gdemonet committed:
WIP - support preparing environments in ducks

gdemonet committed:
WIP - simplify NodeDeployment page

No need to use state to render the job steps.

gdemonet committed:
WIP - update NodeList page

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/rewrite-solution-management` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/rewrite-solution-management`
* Rebase `improvement/rewrite-solution-management` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

gdemonet commented:
Closing as this PR is not actively maintained. The code here should be useful for #2163 efforts (especially on bootstrapping a Python CLI).
